treq/auth.py

```diff
@@ -1,7 +1,144 @@
 from __future__ import absolute_import, division, print_function
+import re
+import time
+import base64
+import hashlib
 
+from six.moves.urllib.parse import urlparse
 from twisted.web.http_headers import Headers
-import base64
+from twisted.python.randbytes import secureRandom
+from requests.utils import parse_dict_header
+
+
+_DIGEST_HEADER_PREFIX_REGEXP = re.compile(r'digest ', flags=re.IGNORECASE)
+
+
+def generate_client_nonce(server_side_nonce):
+    return hashlib.sha1(
+        hashlib.sha1(server_side_nonce).digest() +
+        secureRandom(16) +
+        time.ctime()
+    ).hexdigest()[:16]
+
+
+def _md5_utf_digest(x):
+    if isinstance(x, str):
+        x = x.encode('utf-8')
+    return hashlib.md5(x).hexdigest()
+
+
+def _sha1_utf_digest(x):
+    if isinstance(x, str):
+        x = x.encode('utf-8')
+    return hashlib.sha1(x).hexdigest()
+
+
+def build_digest_authentication_header(agent, **kwargs):
```

Comment:
This shouldn't be public, if it needs to share data with the agent it should probably just be a private method on the agent.